File: scripts/premium_analysis_mirror/ingest_diagnostics.py
"""Ingest diagnostics — mirrors lib/premium-analysis/ingest-diagnostics.ts."""

from __future__ import annotations

from typing import Any

from .mapping import lookup_charge_taxonomy
from .primitives import primary_rollup_date_raw, shipment_package_dedupe_key, to_number


def empty_ingest_diagnostics() -> dict[str, Any]:
    return {
        "duplicateUploadRowsSkipped": 0,
        "duplicateChargeRowsDropped": 0,
        "rowsDroppedCriticalSciCorruption": 0,
        "linesTotal": 0,
        "linesMapped": 0,
        "unmappedSpend": 0.0,
        "shipmentsTotal": 0,
        "shipmentsWithoutTracking": 0,
        "linesMissingShipDate": 0,
        "parseVersions": [],
    }


def _line_is_mapped(rec: dict[str, Any], mapping_lookup: dict[str, dict[str, str]]) -> bool:
    taxonomy = lookup_charge_taxonomy(
        mapping_lookup, rec.get("Carrier Name"), rec.get("Charge Description")
    )
    if not taxonomy:
        return False
    return bool(taxonomy.get("category_1") or taxonomy.get("category_3"))


def _ship_date_present(rec: dict[str, Any]) -> bool:
    raw = primary_rollup_date_raw(rec)
    if not raw or not raw.strip():
        return False
    return not raw.lower().startswith("invoice")


def parse_versions_from_records(records: list[dict[str, Any]]) -> list[str]:
    versions = sorted({str(r.get("_parse_version") or "").strip() for r in records if r.get("_parse_version")})
    return [v for v in versions if v]


def build_ingest_diagnostics(
    records: list[dict[str, Any]],
    base: dict[str, Any],
    mapping_lookup: dict[str, dict[str, str]],
    parse_versions: list[str] | None = None,
) -> dict[str, Any]:
    lines_mapped = 0
    unmapped_spend = 0.0
    lines_missing_ship_date = 0
    shipment_keys: set[str] = set()
    shipments_without_tracking: set[str] = set()

    for rec in records:
        net = to_number(rec.get("Net Amount"))
        if _line_is_mapped(rec, mapping_lookup):
            lines_mapped += 1
        elif net != 0:
            unmapped_spend += net

        if not _ship_date_present(rec):
            lines_missing_ship_date += 1

        ship_key = shipment_package_dedupe_key(rec)
        if not ship_key:
            continue
        shipment_keys.add(ship_key)
        tracking = str(rec.get("Tracking Number") or rec.get("Shipment Reference Number 1") or "").strip()
        if not tracking:
            shipments_without_tracking.add(ship_key)

    versions = parse_versions or parse_versions_from_records(records)

    return {
        **base,
        "linesTotal": len(records),
        "linesMapped": lines_mapped,
        "unmappedSpend": round(unmapped_spend, 2),
        "shipmentsTotal": len(shipment_keys),
        "shipmentsWithoutTracking": len(shipments_without_tracking),
        "linesMissingShipDate": lines_missing_ship_date,
        "parseVersions": versions,
    }

File: lib/premium-analysis/ingest-diagnostics.ts
import { primaryRollupDateRaw, toNumber, type InvoiceRecord } from '@/lib/invoices/csv'
import type { PremiumParseIngestDiagnostics } from '@/lib/premium-analysis/analyze-parse-cache'
import {
  buildChargeDescriptionLookup,
  lookupChargeTaxonomyForRecord,
  shipmentPackageDedupeKey,
} from '@/lib/premium-analysis/analysis-summary'

export type IngestDiagnosticsBase = Pick<
  PremiumParseIngestDiagnostics,
  'duplicateUploadRowsSkipped' | 'duplicateChargeRowsDropped' | 'rowsDroppedCriticalSciCorruption'
>

const ZERO_EXTENDED: Omit<PremiumParseIngestDiagnostics, keyof IngestDiagnosticsBase> = {
  linesTotal: 0,
  linesMapped: 0,
  unmappedSpend: 0,
  shipmentsTotal: 0,
  shipmentsWithoutTracking: 0,
  linesMissingShipDate: 0,
  parseVersions: [],
}

export function emptyIngestDiagnostics(): PremiumParseIngestDiagnostics {
  return {
    duplicateUploadRowsSkipped: 0,
    duplicateChargeRowsDropped: 0,
    rowsDroppedCriticalSciCorruption: 0,
    ...ZERO_EXTENDED,
  }
}

function lineIsMapped(
  rec: InvoiceRecord,
  mappingLookup: ReturnType<typeof buildChargeDescriptionLookup>
): boolean {
  const taxonomy = lookupChargeTaxonomyForRecord(mappingLookup, rec)
  return Boolean(taxonomy?.category_1 || taxonomy?.category_3)
}

function shipDatePresent(rec: InvoiceRecord): boolean {
  const raw = primaryRollupDateRaw(rec)
  return Boolean(raw && raw.trim() && !/^invoice\b/i.test(raw))
}

/**
 * Enrich base ingest counters with mapping, tracking, and spend coverage metrics.
 * Call on the full unfiltered record set after carrier adapters merge.
 */
export function buildIngestDiagnostics(
  records: InvoiceRecord[],
  base: IngestDiagnosticsBase,
  mappingLookup: ReturnType<typeof buildChargeDescriptionLookup>,
  parseVersions: string[] = []
): PremiumParseIngestDiagnostics {
  let linesMapped = 0
  let unmappedSpend = 0
  let linesMissingShipDate = 0

  const shipmentKeys = new Set<string>()
  const shipmentsWithoutTracking = new Set<string>()

  for (const rec of records) {
    const net = toNumber(rec['Net Amount'])
    if (lineIsMapped(rec, mappingLookup)) {
      linesMapped += 1
    } else if (net !== 0) {
      unmappedSpend += net
    }

    if (!shipDatePresent(rec)) linesMissingShipDate += 1

    const shipKey = shipmentPackageDedupeKey(rec)
    if (!shipKey) continue
    shipmentKeys.add(shipKey)
    const tracking = (rec['Tracking Number'] ?? rec['Shipment Reference Number 1'] ?? '').trim()
    if (!tracking) shipmentsWithoutTracking.add(shipKey)
  }

  const versions = [...new Set(parseVersions.map((v) => v.trim()).filter(Boolean))].sort()

  return {
    ...base,
    linesTotal: records.length,
    linesMapped,
    unmappedSpend: +unmappedSpend.toFixed(2),
    shipmentsTotal: shipmentKeys.size,
    shipmentsWithoutTracking: shipmentsWithoutTracking.size,
    linesMissingShipDate,
    parseVersions: versions,
  }
}

export function mergeParseVersions(parts: Array<string | undefined | null>): string[] {
  return [...new Set(parts.map((v) => String(v ?? '').trim()).filter(Boolean))]
}
